src/_guides/language/concurrency/index.md

```diff
@@ -0,0 +1,449 @@
+   the single message that the worker isolate sends.
+
+[`first`]: {{site.dart_api}}/{{site.data.pkg-vers.SDK.channel}}/dart-async/Stream/first.html
+
+The spawned isolate executes the following code:
+
+```dart
+Future _readAndParseJson(SendPort p) async {
+  final fileData = await File(filename).readAsString();
+  final jsonData = jsonDecode(fileData);
+  Isolate.exit(p, jsonData);
+}
+```
+
+The relevant statement is the last one, which exits the isolate,
+sending `jsonData` to the passed-in `SendPort`.
+Message passing between isolates normally involves data copying,
+and thus can be slow and increases with the size of the message —
+O(n) in [big O notation][].
+However, when you send the data using `Isolate.exit()`,
+then the memory that holds the message in the exiting isolate isn’t copied,
+but instead is transferred to the receiving isolate.
+That transfer is quick and completes in constant time — O(1).
+
+[big O notation]: https://en.wikipedia.org/wiki/Big_O_notation
+
+{{site.alert.version-note}}
+  `Isolate.exit()` was added in 2.15.
+  Previous releases support only explicit message passing,
+  using `Isolate.send()` as shown in the next section's example.
+{{site.alert.end}}
+
+The following figure illustrates the communication between
+the main isolate and the worker isolate:
+
+![A figure showing the previous snippets of code running in the main isolate and in the worker isolate](/guides/language/concurrency/images/isolate-api.png)
+
+
+### Sending multiple messages between isolates
+
+If you need more communication between isolates,
+then you need to use the [`send()` method][] of `SendPort`.
+One common pattern, which the following figure shows,
+is for the main isolate to send a request message to the worker isolate,
+which then sends one or more reply messages.
+
+![A figure showing the main isolate spawning the isolate and then sending a request message, which the worker isolate responds to with a reply message; two request-reply cycles are shown](/guides/language/concurrency/images/isolate-custom-bg-worker.png)
+
+For examples of sending multiple messages,
+see the following [isolate samples][]:
+
+* [send_and_receive.dart][],
+  which shows how to send a message from
+  the main isolate to the spawned isolate.
+  It’s otherwise similar to the preceding example.
+* [long_running_isolate.dart][],
+  which shows how to spawn a long-running isolate that
+  receives and sends multiple times.
+
+{% assign samples = "https://github.com/dart-lang/samples/tree/master/isolates" %}
+
+[isolate samples]: {{ samples }}
+[send_and_receive.dart]: {{ samples }}/bin/send_and_receive.dart
+[long_running_isolate.dart]: {{ samples }}/bin/long_running_isolate.dart
+
+{{site.alert.tip}}
+  The `send()` method can be slow.
+  When possible,
+  avoid using `send()` to send large amounts of data.
+  Instead, send the data when the worker isolate exits,
+  as described in the previous section.
+{{site.alert.end}}
+
+## Performance and isolate groups
```

Comment:
For future work, I'm not sure the best way to improve it yet, but I feel the discussion around performance and isolate groups could maybe be better consolidated. Or perhaps isolate groups and their implications could be introduced earlier instead of tacked on in the end. It seems the reader may need to reference back to previous sections to understand this section and modify their previously gained understanding. 

For example, there is talk about the performance characteristics of `Isolate.exit()` earlier, with these clarifications added later on, so it seems the discussion around that is a bit separated.